### PR TITLE
Rework NNCF module placement to track all access scopes of the replaced module

### PR DESCRIPTION
**Pull request by vshampor**
### Changes

When replacing NNCF modules, keep track of all scopes through which the module is accessible in the original model object. When building the DynamicGraph, keep track of the module `id` that the operation is executed in, and reuse this information to detect shared-layer nodes in NNCFGraph. If there are multiple scopes for the module that contains a given operation (represented by the node in the graph), then the node is associated with the lexicographically first scope of the containing module.

### Reason for changes

Some PT models may be written in such a fashion that a given submodule calls a certain other submodule in the model, but at the same time this first submodule does not contain the second in the storage hierarchy (see the test case in the PR for an explicit example of such a model). In this case the storage scope does not coincide with the execution scope of the operations; this also causes problems during module traversal in `replace_modules`.

### Related tickets

75373

### Tests

tests.torch.test_nncf_network.test_wrapping_of_indirect_module_operations

### CODE REVIEW

**Comment by ljaljushkin** [comment]
I wonder how ignored_scope and target_scope will work in case of such model. Am I right, that everything will be in storage-mode now: scopes in original and compressed graphs, scopes in log? User will never know that there's execution-mode scope, will he?

**Comment by vshampor** [comment]
No, in fact, the storage scope comparisons are limited to determining the train-only modules and referencing their scopes against the storage scope that is being tracked during the one-time replacement process with NNCF-wrapped ones. The user would probably have to specify storage scope if they really want to not wrap a part of the model, but otherwise all the other ignored scopes and target scopes are still specified in terms of what is visible in the compressed_graph.dot, i.e. in terms of the execution scope.

**Comment by ljaljushkin** [comment]
> No, in fact, the storage scope comparisons are limited to determining the train-only modules and referencing their scopes against the storage scope that is being tracked during the one-time replacement process with NNCF-wrapped ones. The user would probably have to specify storage scope if they really want to not wrap a part of the model, but otherwise all the other ignored scopes and target scopes are still specified in terms of what is visible in the compressed_graph.dot, i.e. in terms of the execution scope.

It makes sense to document this, it can be not obvious even for us after a while

**Review by alexsu52 — APPROVED** [review]

File: nncf/torch/graph/graph_builder.py [hunk 1]
```diff
@@ -81,14 +78,15 @@ def convert(dynamic_graph: DynamicGraph, input_infos: List[ModelInputInfo] = Non
                 if input_infos[input_id].is_integer_input():
                     is_integer_input = True
 
-            is_shared = layer_name in layer_name_vs_node_counts and layer_name_vs_node_counts[layer_name] > 1
+            is_shared = len(module_id_vs_sorted_scopes_map[dynamic_graph_node.calling_module_id]) > 1
```
[the remark this answers is not in the document]
Comment (vshampor) [reply nncf/torch/graph/graph_builder.py:81]: Consider the following module:
```
class MyModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.module = nn.Conv2d(1, 1, 1)
        self.sequential = nn.Sequential(self.module, self.module)

    def forward(self, x):
        y = self.module(x)
        z = self.sequential(x)
        return y + z
```

There is only one actual module object being called (thrice) in the forward, but the scopes in the dynamic graph for each of the three corresponding `conv2d` nodes will be different. These three nodes will, in essence, "share" the module by the way, in the same way as "shared layers" are understood in TF.

[the remark this answers is not in the document]
Comment (ljaljushkin) [reply nncf/torch/graph/graph_builder.py:81]: Ok, I see. Thank you for clarification.

**Review by ljaljushkin — APPROVED** [review]

**Comment by vshampor** [comment]
SOTA eval PT build 206 is OK

**Pull request merged**